src/main/java/com/bobocode/session/SessionImpl.java

```diff
@@ -0,0 +1,120 @@
+package com.bobocode.session;
+
+import com.bobocode.Dialect;
+import com.bobocode.JDBCRepository;
+import com.bobocode.Transaction;
+import com.bobocode.annotation.Table;
+import com.bobocode.exception.BibernateException;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.sql.DataSource;
+import java.util.List;
+import java.util.Map;
+
+import static com.bobocode.Dialect.SELECT_ALL_BY_PROPERTIES_TEMPLATE;
+import static com.bobocode.Dialect.SELECT_ALL_ID_TEMPLATE;
+import static com.bobocode.Dialect.SELECT_ALL_TEMPLATE;
+import static com.bobocode.Dialect.prepareWhereClause;
+import static com.bobocode.Util.checkIdValidPrimaryKeyType;
+import static com.bobocode.Util.checkNotNullArg;
+import static com.bobocode.Util.checkIsEntity;
+import static com.bobocode.Util.getTableName;
+
+@Slf4j
+public class SessionImpl implements Session {
+    private final Dialect dialect;
+
+    private final JDBCRepository jdbcRepository;
+
+    public SessionImpl(DataSource dataSource, Dialect dialect) {
+        this.dialect = dialect;
+        this.jdbcRepository = new JDBCRepository(dataSource);
+    }
+
+    @Override
+    public <T> T find(Class<T> type, Object primaryKey) {
+        checkNotNullArg(type, "type");
+        checkNotNullArg(primaryKey, "primaryKey");
+        checkIsEntity(type);
+        checkIdValidPrimaryKeyType(type, primaryKey);
+
+        log.trace("Finding {} by id", type.getSimpleName());
+
+        String tableName = getTableName(type);
+        String query = SELECT_ALL_ID_TEMPLATE.formatted(tableName);
+        List<T> foundEntities = jdbcRepository.select(type, query, List.of(primaryKey));
+        if (foundEntities.size() != 1) {
+            throw new BibernateException("More then 1 result were found!");
+        }
+        return foundEntities.get(0);
+    }
+
+    @Override
+    public <T> List<T> findAll(Class<T> type, Long limit, Long offset) {
+        checkNotNullArg(type, "type");
+        checkNotNullArg(limit, "limit");
+        checkIsEntity(type);
+
+        log.trace("Finding all {}", type.getSimpleName());
+
+        String tableName = getTableName(type);
+        String query = SELECT_ALL_TEMPLATE.formatted(tableName) + dialect.getLimitOffsetClause(limit, offset);
+        List<Object> properties = offset != null ? List.of(limit, offset) : List.of(limit);
+
+        return jdbcRepository.select(type, query, properties);
+    }
+
+    @Override
+    public <T> List<T> findAll(Class<T> type, Map<String, Object> properties) {
+        checkNotNullArg(type, "type");
+        checkNotNullArg(properties, "properties");
+        checkIsEntity(type);
+
+        log.trace("Finding {} by properties", type.getSimpleName());
+
+        String tableName = getTableName(type);
+        String query = SELECT_ALL_BY_PROPERTIES_TEMPLATE.formatted(tableName, prepareWhereClause(properties));
+
+        List<Object> values = properties.entrySet().stream()
+                .sorted(Map.Entry.comparingByKey())
```

Comment:
why do we sort them?

Reply:
@DanielChanger We have map with name and value of columns (or properties). Key (sorted) are used for designing "where" clause (Dialect.prepareWhereClause(Map<String, Object> properties)). Then value are pass to PreparedStatement as parameters. I use sorting by property name (keys) to matching name of property to its value.

Reply:
Ah, seems like I got it. I thought that maybe we should be passing map of column names and values